### DISCUSSION TITLE
ASP.NET core, Scheduling jobs from services

### DISCUSSION THREAD

**Discussion by Primital** [quartznet/quartznet#1241] [Q&A]
I am working in a project where users will be able to create reminders for stuff in our product.
I am having difficulty achieving this type of dynamic scheduling with quartz, and am in need of some help.

In ConfigureServices:
```
            var scheduler = StdSchedulerFactory.GetDefaultScheduler().GetAwaiter().GetResult();
            services.AddSingleton(scheduler);
            scheduler.Start();
```

The flowchart looks like this:
1. User posts new reminder to controller.
2. Controller calls service to create model in database.
3. Service has scheduler injected from ConfigureServices
4. In service, JobDetail is created from predefined job with some JobData from the user request
5. Trigger is created based on user reminder data (start date, repeat count, etc)
6. Before returning created reminder object, scheduler is called with `scheduler.ScheduleJob(jobDetail, trigger)`

Using the debugger I found that the call to `ScheduleJob` returns a `Task` with `Status` `RanToCompletion` regardless if the trigger is set to start in the future.
Nothing is found in the logs at the trigger date, nor any following dates.

The scheduler has `IsStarted: true`

It seems all documentation is for predefined tasks on a predefined schedule. Is it not possible to add tasks during runtime?

The JobDetail is defined like this:
```
JobBuilder jobBuilder = JobBuilder.Create<ReminderJob>()
    .WithIdentity("reminderJob" + reminder.Id.ToString(), "reminderJobs")
    .UsingJobData("userId", reminder.UserId)
    .UsingJobData("reminderText", reminder.Notes);
return jobBuilder.Build();
```
The Trigger is defined like this:
```
var triggerBuilder = TriggerBuilder.Create()
    .WithIdentity("reminderTrigger" + reminder.Id.ToString(), "reminderTriggers")
    .StartAt(reminder.ReminderDate)
    .WithSimpleSchedule(x => x
    .WithIntervalInMinutes(reminder.IntervalLength)
    .RepeatForever());
return triggerBuilder.Build();
```
The ReminderJob is defined like this:
```
public class ReminderJob : IJob
{
    private readonly ILogger<ReminderJob> _logger;
    public ReminderJob(ILogger<ReminderJob> logger)
    {
        _logger = logger;
    }
    public async Task Execute(IJobExecutionContext context)
    {
        try
        {
            _logger.LogInformation("This should log but does not");
            await Task.CompletedTask;
        }
        catch (Exception e)
        {
            throw new JobExecutionException(e, true);
        }
    }
}
```

**Comment by asulwer** [comment]
is this an issue with the code?  this should be a discussion instead.

i load a cron expression from my db at runtime to create the jobs

**Reply by Primital** [reply]
As far as I can tell this is a discussion thread?

The job is defined like this:
```
            JobBuilder jobBuilder = JobBuilder.Create<ReminderJob>()
                .WithIdentity("reminderJob" + reminder.Id.ToString(), "reminderJobs")
                .UsingJobData("userId", reminder.UserId)
                .UsingJobData("reminderText", reminder.Notes);
            return jobBuilder.Build();
```

The trigger is defined like this:
```
var triggerBuilder = TriggerBuilder.Create()
    .WithIdentity("reminderTrigger" + reminder.Id.ToString(), "reminderTriggers")
    .StartAt(reminder.ReminderDate)
    .WithSimpleSchedule(x => x
    .WithIntervalInMinutes(reminder.IntervalLength)
    .RepeatForever());
return triggerBuilder.Build();
```

The ReminderJob is defined like this:
```
    public class ReminderJob : IJob
    {
        private readonly ILogger<ReminderJob> _logger;
        public ReminderJob(ILogger<ReminderJob> logger)
        {
            _logger = logger;
        }
        public async Task Execute(IJobExecutionContext context)
        {
            try
            {
                _logger.LogInformation("This should log but does not");
                await Task.CompletedTask;
            }
            catch (Exception e)
            {
                throw new JobExecutionException(e, true);
            }
        }
    }

```

**Reply by asulwer** [reply]
you are correct it is a discussion.  sorry about that.

i cannot see anything, with just a casual glance, wrong with this code.  the only difference with mine and yours is your trigger.

```
ITrigger trigger = TriggerBuilder
                            .Create()
                            .WithIdentity($"{site.Name}", $"{schedule.Id}")
                            .WithCronSchedule(schedule.Frequency, cron => cron
                                .WithMisfireHandlingInstructionDoNothing())
                            .ForJob(job)
                            .Build();
```

the only real difference is that i assign the trigger to the job ForJob.  maybe you are doing this later?
